lib/translator/inquiry.cc

```diff
@@ -0,0 +1,247 @@
+
+#include "inquiry.h"
+#include "string.h"
+
+namespace translator {
+
+// Creates and validates a Inquiry Command struct
+StatusCode RawToScsiCommand(absl::Span<const uint8_t> raw_cmd,
+                            scsi_defs::InquiryCommand &cmd) {
+  if (raw_cmd.empty()) {
+    DebugLog("buffer is empty or nullptr");
+    return StatusCode::kInvalidInput;
+  }
+
+  uint8_t raw_opcode = raw_cmd[0];
+  if (raw_opcode > 0xaf) {
+    DebugLog("raw opcode is out of range: %x", raw_opcode);
+    return StatusCode::kInvalidInput;
+  }
+
+  scsi_defs::OpCode opcode = static_cast<scsi_defs::OpCode>(raw_opcode);
+
+  if (opcode != scsi_defs::OpCode::kInquiry) {
+    absl::string_view expected_cmd_str =
+        ScsiOpcodeToString(scsi_defs::OpCode::kInquiry);
+    absl::string_view cmd_str = ScsiOpcodeToString(opcode);
+    DebugLog("wrong opcode. expected %s got %s.", expected_cmd_str, cmd_str);
+    return StatusCode::kInvalidInput;
+  }
+
+  // TODO: check invalid parameters in scsi_command
+  cmd = *reinterpret_cast<const scsi_defs::InquiryCommand *>(&raw_cmd[1]);
+  return StatusCode::kSuccess;
+}
+
+void TranslateStandardInquiry(
+    const nvme_defs::IdentifyControllerData &identify_controller_data,
+    const nvme_defs::IdentifyNamespace &identify_namespace_data,
+    absl::Span<uint8_t> buffer) {
+  scsi_defs::InquiryData result = {
+      .version = scsi_defs::Version::kSpc4,
+      .response_data_format = scsi_defs::ResponseDataFormat::kCompliant,
+      .additional_length = 0x1f,
+      .tpgs = scsi_defs::TPGS::kNotSupported,
+      .protect = (identify_namespace_data.dps.pit == 0 &&
+                  identify_namespace_data.dps.md_start == 0)
+                     ? 0
+                     : 1,
+      .cmdque = 1};
+
+  // Shall be set to “NVMe" followed by 4 spaces: “NVMe    “
+  // Vendor Identification is not null terminated.
+  static_assert(sizeof(result.vendor_identification) ==
+                kNvmeVendorIdentification.size());
+
+  memcpy(result.vendor_identification, kNvmeVendorIdentification.data(),
+         kNvmeVendorIdentification.size());
+
+  // Shall be set to the first 16 bytes of the Model Number (MN) field within
+  // the Identify Controller Data Structure
+  memcpy(result.product_identification, identify_controller_data.mn,
+         sizeof(result.product_identification));
+
+  /*
+  Shall be set to the last 4 ASCII graphic characters in the range of 21h-7Eh
+  (i.e. last 4 non-space characters) of the Firmware Revision (FR) field within
+  the Identify Controller Data Structure.
+  */
+  int idx = 3;
+  for (int i = 7; i >= 0; i--) {
+    if (identify_controller_data.fr[i] != ' ' &&
+        identify_controller_data.fr[i] >= 0x21 &&
+        identify_controller_data.fr[i] <= 0x7e) {
+      result.product_revision_level[idx] = identify_controller_data.fr[i];
+      if (idx-- == 0) break;
+    }
+  }
+
+  memcpy(buffer.data(), &result, sizeof(result));
+}
+
+void TranslateSupportedVpdPages(absl::Span<uint8_t> buf) {
```

Comment:
nit: keep naming consistent. Since we are using `buffer` elsewhere, lets change `buf` -> `buffer`